simple_network_sim/inference.py

```diff
@@ -0,0 +1,757 @@
+        return InferredInfectionProbability(value, mean, shape, kernel_sigma, fitter.random_state)
+
+    def generate_perturbated(self) -> InferredInfectionProbability:
+        """ From current parameter, add a perturbation to infection probability and return
+        a newly created perturbated parameter:
+
+        .. math::
+            P_t^* \sim K(P_t | P_{t-1}) \sim Uniform(max(P_{t-1} - \sigma, 0),\min(\sigma + P_{t-1}, 1))
+
+        A uniform perturbation of range 2 * kernel_sigma is targeted, with a floor at 0 so
+        that the infection probability remains valid. This is correct as the algorithm
+        states that the particle should be re-sampled as long as the perturbation brings it
+        out of bounds, and the truncation of a uniform distribution is still a uniform
+        distribution.
+
+        :return: New parameter which is similar to self up to a perturbation
+        """
+        sigma = self.kernel_sigma
+        value = self.value.copy()
+        value.Value = self.random_state.uniform(np.maximum(value.Value - sigma, 0), np.minimum(value.Value + sigma, 1.))
+        return InferredInfectionProbability(value, self.mean, self.shape, self.kernel_sigma, self.random_state)
+
+    def validate(self) -> bool:
+        """ Checks that the particle is valid, i.e. that infection probability is
+        between 0 and 1.
+
+        :return: Whether the parameter is valid
+        """
+        return np.all(self.value.Value > 0.) and np.all(self.value.Value < 1.)
+
+    def prior_pdf(self) -> np.ndarray:
+        """ Compute pdf of the prior distribution evaluated at the parameter x.
+        Infection probability has a prior a beta distribution. The pdf is evaluated
+        at the current value of the parameter.
+
+        :return: pdf value of prior distribution evaluated at x
+        """
+        return np.prod(stats.beta.pdf(self.value.Value, self.shape,
+                                      self.shape * (1 - self.mean.Value) / self.mean.Value))
+
+    def perturbation_pdf(self, x: pd.DataFrame) -> np.ndarray:
+        """ Compute pdf of the perturbation evaluated at the parameter x,
+        from the current parameter. In ABC-SMC when a particle is sampled
+        from the previous population it is slightly perturbed:
+
+        .. math::
+            P_t^* \sim K(P_t | P_{t-1}) \sim Uniform(max(P_{t-1} - \sigma, 0),\min(\sigma + P_{t-1}, 1))
+
+        Given in the sampling the perturbation was capped and floored in [0, 1]
+        to keep the particle valid, it results in a truncated uniform
+        distribution which is reflected in the pdf.
+
+        :param x: Particle to evaluate the pdf at
+        :return: pdf value of perturbation from previous particle evaluated at x
+        """
+        return np.prod(uniform_pdf(x.Value, np.maximum(self.value.Value - self.kernel_sigma, 0),
+                                   np.minimum(self.value.Value + self.kernel_sigma, 1)))
```

Comment:
Going back to my previous comment, I'm creating a new one since that one became outdated. This is not a big deal, but technically both the dataframe passed to this function and the one passed to the same function in InferredInitialInfections should have the same columns. This is something the type system would complain if it knew about it (and possibly require a cast or just deny)

Reply:
Thanks, this is indeed True for infection proba and intial infection, but is not true anymore since I added contact multiplier, of which the dataframe has a different shape. So unfortunately we can't be stricter than currently